rust/blockstore/src/arrow/block/types.rs

```diff
@@ -158,36 +169,103 @@ impl Block {
             // SAFETY: the call is made safe by the following inconstants:
             // - `mid >= 0`: by definition
             // - `mid < size`: `mid = size / 2 + size / 4 + size / 8 ...`
-            let mut cmp = prefix_array.value(mid).cmp(prefix);
-
-            // Continue to compare the key if prefix matches
-            if let (Equal, Some(k)) = (cmp, key) {
-                cmp = K::get(self.data.column(1), mid)
-                    // Key type do not have total order because of floating point values
-                    // But in our case NaN should not be allowed so we should always have total order
-                    .partial_cmp(k)
-                    .expect("Array values should be comparable.");
-            }
-
-            base = if cmp == Less { mid } else { base };
+            let prefix = prefix_array.value(mid);
+            let key = K::get(self.data.column(1), mid);
+            let cmp = f((prefix, key));
```

Comment:
I thought about that but didn't see a clean way to do it. Do you have ideas?

Reply:
Maybe we can ask the comparator to take in a key handle:
- Before: `F: FnMut((&'me str, K)) -> Ordering`
- After: `F: FnMut(&'me str, C) -> Ordering, C: Fn() -> &'me K`

And we pass in `|| K::get(...)` to the comparator.

Alternatively, ask the caller to pass in two comparators, one for prefix and the other for key. The second comparator is invoked only when the first yields `Ordering::Equal` (basically using a `then_with`)

Maybe there are better ways to do this, but this is what I have in mind right now.

Reply:
yeah, that works

Reply:
this turns out to be somewhat difficult:

> `F: FnMut(&'me str, C) -> Ordering, C: Fn() -> &'me K`:

Can't work without Boxing the key fetch callback.

> Alternatively, ask the caller to pass in two comparators, one for prefix and the other for key. The second comparator is invoked only when the first yields Ordering::Equal (basically using a then_with)

Makes life annoying for callers because the key cmp function parameter must be optional. But if `None` is provided, Rust complains that that the parameter type is unknown.

------------------------

Are you ok leaving as-is for now and revisiting if it shows up in flamegraphs?

Reply:
I'm totally fine with the current impl unless there is evidence showing that fetching the key will significantly slow us down